src/Functions/FunctionsConversion.cpp

```diff
@@ -1573,6 +1574,55 @@ struct ConvertImpl
                         arguments, result_type, input_rows_count, additions);
             }
         }
+        else if constexpr (std::is_same_v<FromDataType, DataTypeInterval> && std::is_same_v<ToDataType, DataTypeInterval>)
+        {
+            IntervalKind to = typeid_cast<const DataTypeInterval *>(result_type.get())->getKind();
+            IntervalKind from = typeid_cast<const DataTypeInterval *>(arguments[0].type.get())->getKind();
+
+            if (from == to || arguments[0].column->empty())
+                return arguments[0].column;
+
+            const auto &map = getGranularityMap();
+            Int64 conversion_factor = 1;
+            Int64 result_value;
+
+            int from_position = map.at(from).first;
+            int to_position = map.at(to).first; // Positions of each interval according to granurality map
+
+            if (from_position < to_position)
+            {
+                for (int i = from_position - 1; i <= to_position; ++i)
+                {
+                    // Find the kind that matches this position
+                    for (const auto &entry : map)
+                    {
+                        if (entry.second.first == i)
+                        {
+                            conversion_factor *= entry.second.second;
+                            break;
+                        }
+                    }
+                }
```

Comment:
I think this would become easier if you would use the `IntervalKind` to find the necessary entries.